Improve `reload_scripts` in `ScriptEditor`

Rindbee opened the pull request:
This PR is based on #66916 and requires #66880. It also uses the signal in #66880 to work.

#66880 try to use a signal to update the internal editor. Because for external editing via LSP, not all scripts in the list need to be updated, and `reload` may be called multiple times per script. Apply #66880, which should solve the problem with LSP.

#66916 is mainly for the tool script, as the reload of the tool script is a soft reload.

**Note:**
    
Currently, the script list in `ScriptEditor` is used to detect changed script files. In this way, those scripts that are not opened in the script list will be missed, so that these scripts cannot be reloaded and updated when the editor is refocused.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

Rindbee committed:
Improve `reload_scripts` in `ScriptEditor`

1. Outdated script documents are cleaned up and updated with the latest script
documents.
2. Tool script reloads are now taken into account.

**Note:**

Currently, the script list in `ScriptEditor` is used to detect changed script
files. In this way, those scripts that are not opened in the script list will
be missed, so that these scripts cannot be reloaded and updated when the editor
is refocused.

akien-mga commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -49,6 +49,7 @@
 #include "editor/find_in_files.h"
 #include "editor/node_dock.h"
 #include "editor/plugins/shader_editor_plugin.h"
+#include "modules/gdscript/gdscript.h"
```
This file should stay generic and can't depend on a specific module. This would make it impossible to build Godot with `module_gdscript_enabled=no`.

If GDScript specific behavior is needed here, it needs to go through the Script/ScriptLanguage abstraction and be implemented in the GDScript module.

Rindbee replied:
OK, thanks for pointing it out.

Rindbee committed:
Improve `load_source_code`/`load_text` for `Script`/`TextFile`

Set modification time when successfully loaded.

`load_source_code` has been implemented in `GDScript`/`CSharpScript`. It
is important for the reload of tool scripts. Now abstracted in `Script`.

YuriSizov commented:
Both linked issues are now fixed and closed. Is this still relevant?

HolonProduction commented:
I'm kinda confused by this PR. The issues originally linked are already closed. What is this PR achieving?

Rindbee commented:
It seems there was a similar unresolved issue at the time, but I forgot to make a record of it. 
It hasn't been maintained for a long time, so I'm not sure what the current situation is.
